Provide `add_task_number` pre-commit hook

vanya909 opened the pull request:
This hook will automatically insert task number to commit message. It will retrieve task number from current branch name if possible.

Note: You can test this hook out, I've applied it in this branch. Just run `pre-commit install --hook-type commit-msg` and try to commit something on this branch, the task number will be appended automatically.

Task: SA3P-123

vanya909 committed:
Provide `add_task_number` pre-commit hook

This hook will automatically insert task number to commit message.
It will retrieve task number from current branch name if possible.

Task: SA3P-123

ron8mcr commented on README.md:
```diff
@@ -90,3 +90,22 @@ Please note that by default there are always ignored errors with these keywords:
 
 This is it!
+
+### `add_task_number`
```
Could you please add example of what should be added to `.pre-commit-config` to make this hook work? Including default parameters for regex and format

vanya909 committed:
Fix README and hook stage

README was fixed, also was provided correct hook stage for new
pre-commit hook.

Task: SA3P-123

ron8mcr commented on README.md:
```diff
@@ -90,3 +90,35 @@ Please note that by default there are always ignored errors with these keywords:
+    hooks:
+      - id: add-task-number
+        args:
+        - "--regex=feature/(?P<task>[A-Z]{2,11}-[0-9]{1,6})-.*"
```
`regex` is too common name, I suggest to use `branch-regex`

ron8mcr commented on pre_commit_hooks/add_task_number/util.py:
```diff
@@ -0,0 +1,109 @@
+        stripped = line.strip().lower()
+
+        # Skip empty lines and comment lines
+        if stripped == "" or stripped.startswith("#"):
```
Comment char may be changed, let's retrieve it from git config
https://git-scm.com/docs/git-config#Documentation/git-config.txt-corecommentString

vanya909 replied:
I've checked the case, looks like I can avoid deleting of comments, so I think it's better to remove this behavior in order to not add more logic related to git.

ron8mcr commented on pre_commit_hooks/add_task_number/util.py:
```diff
@@ -0,0 +1,109 @@
+    return False
+
+
+def get_message_without_comment_section(message: str) -> str:
```
```suggestion
def strip_comment_section(message: str) -> str:
```

vanya909 committed:
Update regex name

Now it is called branch-regex.

Task: SA3P-123

dmitry-mightydevops commented:
@kseniyashaydurova pls review and decide today how to proceed with repo/repos

vanya909 committed:
Rename util function

Rename `get_message_without_comment_section` to `strip_comment_section`.

Task: SA3P-123

kseniyashaydurova commented on pre_commit_hooks/add_task_number/cli.py:
```diff
@@ -0,0 +1,42 @@
+        description="Script that adds task number to each commit",
+    )
+
+    parser.add_argument("filenames", nargs="+")
```
LEt's add help text for each param

kseniyashaydurova commented on pre_commit_hooks/add_task_number/util.py:
Let's name this file `main.py`? Cause basically here is the main pre-commit hook logic is located

kseniyashaydurova commented on pre_commit_hooks/add_task_number/util.py:
```diff
@@ -0,0 +1,109 @@
+        if stripped == "" or stripped.startswith("#"):
+            continue
+
+        if task.lower() in stripped:
```
Let's replace to this? Seems clear and reduces lines
```
retrun task.lower() in stripped
```

vanya909 replied:
Unfortunately, we cannot do so, this check is located in the loop, so by making instant return we will force this code to check only the first line every time, because it will make return after first iteration

kseniyashaydurova replied:
Ah sorry, right

vanya909 committed:
Remove deleting of comments in commit message

Comments deleting is redundant and requires adding of new logic to get
value from git config, as comment string may differ depending on it.

Task: SA3P-123

kseniyashaydurova commented on pre_commit_hooks/add_task_number/util.py:
```diff
@@ -0,0 +1,100 @@
+GIT_COMMENT_SECTION_LINE = r"#?[ ]*-+ >8 -+"
+
+
+def get_current_branch() -> str:
```
LEt's move it to common util.py file? We have there all git related commands that can be resued in different hooks

kseniyashaydurova commented on pre_commit_hooks/add_task_number/util.py:
```diff
@@ -0,0 +1,100 @@
+    return task_number
+
+
+def is_task_in_message(contents: str, task: str) -> bool:
```
I suppose this function should check whther task presents in a desired format? I.e. this should be considered as the task presents? 

```
blblallslfsldfsf ABC-123
```

I mean that only lines in a passed task `format` should be considered as added, right?

kseniyashaydurova commented on tests/test_add_task_number.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def test_get_message_without_comment_section():
+    """Test that comment section will be cut of by this util function."""
```
```suggestion
    """Test that comment section will be cut off by this util function."""
```

kseniyashaydurova commented on tests/test_add_task_number.py:
```diff
@@ -0,0 +1,86 @@
+        assert last_commit_message == expected_commit_message
+
+
+def test_empty_commit_message_will_not_be_modified(temp_git_dir):
```
LEt's add a test for behavior:
- we use this pre-commit hook
- our branch doesn't have any <task> in its name

kseniyashaydurova commented on pre_commit_hooks/add_task_number/cli.py:
```diff
@@ -0,0 +1,42 @@
+    parser.add_argument("filenames", nargs="+")
+    parser.add_argument(
+        "--branch-regex",
+        default=r"(feature|fix)/(?P<task>[A-Z0-9]+-[0-9]+)-.*",
```
```suggestion
        default=r"^(feature|fix)/(?P<task>[A-Z0-9]+-[0-9]+)-.*",
```

kseniyashaydurova commented on README.md:
```diff
@@ -90,3 +90,35 @@ Please note that by default there are always ignored errors with these keywords:
+repos:
+  - repo: https://github.com/saritasa-nest/saritasa-pre-commit-hooks
+    rev: 0.0.2
+    hooks:
```
LEt's specify stages in the example to run it only on commit message stage?
![image](https://github.com/user-attachments/assets/fd694360-26dc-49eb-87fe-6e22eadf78a3)

kseniyashaydurova replied:
And fix indernts for `args`

vanya909 replied:
I can add it, but I've added `commit-msg` stage to `.pre-commit-hooks.yaml` and this stage will be used by default, do we need to add it manually?

kseniyashaydurova replied:
Ah, ok , then nevermind

kseniyashaydurova commented on pre_commit_hooks/add_task_number/cli.py:
```diff
@@ -0,0 +1,42 @@
+    return parser.parse_args(argv)
+
+
+def main(argv=None):
```
LEt's output some info whether `Task` was autoadded or not? and which task was added?

vanya909 replied:
I will add it but looks like the `pre-commit` suppresses output until some error happens.
We can set `verbose: true` for this exact pre-commit hook, but it is considered to be a [debug function](https://stackoverflow.com/a/72898524)

kseniyashaydurova reviewed (changes requested):
@dmitry-mightydevops I think we can use this workflow as a common accepted one, so it works like this for us:
- in case we have `feature/ABC-123` branch format - task id will be auto added in each commit on each `git commit` cmd run
- in case if our branch doesn't fir this format - no problem, task is just not added from branch name and our `jira-pre-commit` hook validates whther the task is absent or not

vanya909 committed:
Add help text to CLI params

Help test was added to all CLI params.

Task: SA3P-123

vanya909 committed:
Rename util

This file was renamed to `main`.

Task: SA3P-123

vanya909 committed:
Fix remarks from code review

Fix small remarks.

Task: SA3P-123

kseniyashaydurova reviewed (changes requested):
Pls process one more scenario:
![image](https://github.com/user-attachments/assets/25efef57-1cb4-4c25-8286-d15dcf608a57)

I think it should behave in a similar way as it does in such a case:
![image](https://github.com/user-attachments/assets/47191bf0-db17-4784-9760-3e8ece86b10c)

vanya909 committed:
Fix checking for task number in commit message

Now will be checked that the task in message is a whole word, not a part
of it.

Task: SA3P-123

vanya909 committed:
Remove `stages` from `README.md`

`.pre-commit-hook.yaml` already has default stages.

Task: SA3P-123

kseniyashaydurova commented on pre_commit_hooks/add_task_number/cli.py:
```diff
@@ -0,0 +1,54 @@
+    )
+
+    parser.add_argument(
+        "filenames",
```
```suggestion
        "commit_msg",
```

vanya909 committed:
Fix args for CLI

Now CLI accepts only one file name which is commit message file.

Task: SA3P-123
